src/libs/protoboard/blackboard_manager.h

```
@@ -0,0 +1,331 @@
+#include <aspect/logging.h>
+#include <blackboard/utils/on_message_waker.h>
+#include <core/threading/thread.h>
+#include <interfaces/ProtobufPeerInterface.h>
+
+#include <boost/fusion/include/any.hpp>
+#include <boost/fusion/include/for_each.hpp>
+#include <boost/fusion/include/std_tuple.hpp>
+#include <type_traits>
+#include <unordered_map>
+#include <vector>
+
+namespace protoboard {
+
+/** Map a blackboard interface type to a blackboard interface ID.
+ * Must be implemented by the user. Will be called for every type used with
+ * an @a bb_iface_manager.
+ * @return The interface name (ID) that should be used for the given @tparam type */
+template <class IfaceT>
+std::string iface_id_for_type();
+
+/** Must be implemented by the user.
+ * @return a vector of paths where ProtoBuf should look for message definitions */
+std::vector<std::string> proto_dirs();
+
+/**
+ * Container for an opened interface of type @tparam IfaceT.
+ * @tparam MessageTypeList must be a @a type_list of the message types that should be
+ * handled on @tparam IfaceT.
+ */
+template <class IfaceT, class MessageTypeList>
+class bb_iface_manager
+{
+public:
+	/// Constructor. Not responsible for actual initialization.
+	bb_iface_manager() : interface_(nullptr), blackboard_(nullptr), waker_(nullptr)
+	{
+	}
+
+	/** Open an interface of the given type with the ID supplied by @a iface_id_for_type and
+	 * register to wake the given thread when any of the given types arrives.
+	 * @param blackboard The blackboard to use
+	 * @param thread The thread to wake */
+	void
+	init(fawkes::BlackBoard *blackboard, fawkes::Thread *thread)
+	{
+		blackboard_ = blackboard;
+		interface_  = blackboard_->open_for_writing<IfaceT>(iface_id_for_type<IfaceT>().c_str());
+		waker_      = new fawkes::BlackBoardOnMessageWaker(blackboard, interface_, thread);
+	}
+
+	/// Cleanup.
+	void
+	finalize()
+	{
+		if (waker_)
+			delete waker_;
+		if (blackboard_ && interface_)
+			blackboard_->close(interface_);
```

I'd suggest removing the conditions and `delete`/`close` unconditionally. If those are `null`, something is broken anyway, so let it crash so it's obvious?

---

Sure, the first `if` is unnecessary because `delete nullptr` is well-defined. See https://en.cppreference.com/w/cpp/language/delete
But I'd rather leave the second one because that can actually trigger unnecessary null pointer dereferences. Not sure if the condition actually arise in practice though.